pkg/actuators/machine/machine.go

```diff
@@ -0,0 +1,321 @@
+	"github.com/openshift/cluster-api-provider-ovirt/pkg/utils"
+	machinev1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	apierrors "github.com/openshift/machine-api-operator/pkg/controller/machine"
+	"github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned/typed/machine/v1beta1"
+	ovirtsdk "github.com/ovirt/go-ovirt"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	InstanceStatusAnnotationKey = "machine.openshift.io/instance-state"
+	ErrorInvalidMachineObject   = "error validating machine object fields"
+	userDataSecretKey           = "userData"
+)
+
+type machineScope struct {
+	context.Context
+	ovirtClient         ovirtC.OvirtClient
+	client              client.Client
+	osClient            osclientset.Interface
+	machine             *machinev1.Machine
+	machinesClient      v1beta1.MachineV1beta1Interface
+	machineProviderSpec *ovirtconfigv1.OvirtMachineProviderSpec
+}
+
+func newMachineScope(
+	ctx context.Context,
+	ovirtClient ovirtC.OvirtClient,
+	client client.Client,
+	machinesClient v1beta1.MachineV1beta1Interface,
+	machine *machinev1.Machine,
+	providerSpec *ovirtconfigv1.OvirtMachineProviderSpec) *machineScope {
+
+	config := ctrl.GetConfigOrDie()
+	osClient := osclientset.NewForConfigOrDie(rest.AddUserAgent(config, "cluster-api-provider-ovirt"))
+	return &machineScope{
+		Context:             ctx,
+		ovirtClient:         ovirtClient,
+		client:              client,
+		machinesClient:      machinesClient,
+		machine:             machine,
+		machineProviderSpec: providerSpec,
+		osClient:            osClient,
+	}
+}
+
+// create creates machine if it does not exists.
+func (ms *machineScope) create() error {
+	if vErr := validateMachine(ms.ovirtClient, ms.machineProviderSpec); vErr != nil {
+		return vErr
+	}
+	// creating a new instance, we don't have the vm id yet
+	instance, err := ms.ovirtClient.GetVmByName(ms.machine.Name)
+	if err != nil {
+		return err
+	}
+	// TODO: Handle that case in the actuator and make sure to return or at least check the impact on patch machine
+	if instance != nil {
+		klog.Infof("Skipped creating a VM that already exists.\n")
+		return nil
+	}
+
+	ignition, err := ms.getIgnition()
+	if err != nil {
+		return apierrors.CreateMachine("Error getting VM ignition: %v", err)
+	}
+
+	instance, err = ms.ovirtClient.CreateVmByMachine(ms.machine, ignition, ms.machineProviderSpec)
+	if err != nil {
+		return fmt.Errorf("Error creating Ovirt instance: %v", err)
+	}
+
+	// Wait till ready
+	// TODO: export to a regular function
+	// TODO: Why are we always setting error to nil?!
+	err = util.PollImmediate(RetryIntervalInstanceStatus, TimeoutInstanceCreate, func() (bool, error) {
+		instance, err := ms.ovirtClient.GetVmByMachine(*ms.machine)
+		if err != nil {
+			return false, nil
```

Comment:
Is this correct? How will the `util.PollImmediate` ever return an error if there is never an error returned inside?

Reply:
haha asked the same question on the todo haha

Reply:
Oops...